texar/core/layers.py

```diff
@@ -664,6 +664,14 @@ def forward(self, input: torch.Tensor) -> torch.Tensor:  # type: ignore
                 layer_output = layer(input)
                 layer_outputs.append(layer_output)
 
+        # the merge dimension cannot be determined until we get the output from
+        # individual layers.
+        # In case of reduce pooling operations, feature dim is removed and
+        # channel dim is merged.
+        # In non-reduce pooling operations, feature dim is merged.
+        if self._dim is None:
+            self._dim = layer_outputs[0].dim() - 1
+
         if self._mode == 'concat':
             outputs = torch.cat(tensors=layer_outputs, dim=self._dim)
         elif self._mode == 'elemwise_sum':
```

Comment:
(This comment is for lines 685-708.)

The logic here is almost the same, maybe we can unify them by e.g. creating a class-level dictionary mapping modes to torch functions and just looking them up.

Reply:
True, I agree with that. Removes a lot of repetition. This was copied with little to no changes from `texar-TF`.